feat: SSB-2 add endpoints for password recovery and validate password…

AntonioLok opened the pull request:
# Description
This PR adds 2 new endpoints which will be used for to reset a user password if they have forgot it.

1-  1st endpoint: POST endpoint `users/forgot-password` this endpoint sends a recovery email with a link to reset a user password. This link will contain a hash/token that is created and stored in our DB for a specific user. If the user requesting to reset his/her password does not exist, we will return a 404. Otherwise if it exists, we will return a 200 and an email will be sent to the user. In the body, we should send the username.

2-  2nd endpoint: POST endpoint `users/reset-password` this endpoint serves to validate the token that is passed in the request parameter and update the user password. In the body, we should send the token, the username and the new password.

 We check against the token (if there is any) and the expiration time that we have stored for the user requesting a password recovery. If the token is valid,  user should be able to update their new password in the DB. If token is not valid, return a 401

# Test
0 - checkout this branch `git checkout SSB-2-create-endpoints-for-password-recovery-and-validate-link-hash`
1 - Install dependencies `npm i`
2 - Set the required environment variables `JWT_SECRET`, `NODEMAILER_TRANSPORTER_USERNAME`, `NODEMAILER_TRANSPORTER_PASSWORD` by creating a `.env` file in the main directory and set your secret in that file ie. `JWT_SECRET=HELLO`. For the email address and email password, please use: 
```
NODEMAILER_TRANSPORTER_USERNAME=os.recovery.pw@gmail.com
NODEMAILER_TRANSPORTER_PASSWORD=random.password
```

The email address and password is used by the nodemailer module to send emails to other users with the recovery password link.
3 - Run the app using `npm run start`.
4 - Now, using the terminal, send the following HTTP requests to test our 2 new endpoints:

To test our endpoints, we need an account first (SKIP this if you have an account):
- Endpoint for registering an acc `/users/register`:
    -  method: POST
    -  body parameters:
        - `username` type: string
        - `password` type: string

sample command:
`curl -d "username=antonio@gmail.com&password=password" -X POST http://localhost:8000/users/register`.

The response should be:

```
 {"statusCode":201,"statusText":"CREATED","data":null}
```

Now that you have your email, we can do our test!

-  1st endpoint `/users/forgot-password`
    -  method: POST 
    -  body parameters:
        - `username` type: string

sample command:
`curl -d "username=antoniochan030@gmail.com" -X POST http://localhost:8000/users/forgot-password`

Plz do not use the above sample command. Replace the username with the email you used to register before. Note that an email will be sent to that email address so if you were to try the above sample command, I will be getting the email on my personal email.

The response should be:

```
 {"statusCode":200,"statusText":"SUCCESS","data":null}
```
Now try with an email that does not exist ie. username=antoniochan04324324234cz30@gmail.com
`curl -d "username=antoniochan04324324234cz30@gmail.com" -X POST http://localhost:8000/users/forgot-password`

The response should be :

```
{"statusCode":400,"statusText":"BAD_REQUEST","message":"The request was malformed or missing a required parameter."}
```

-  2nd endpoint `/users/reset-password`
    -  body parameters:
    - `username` type: string
    - `token` type: string
    - `newPassword` type: string


Check your inbox. An email should have been sent to you with the password reset link. It should look something like this: `http://localhost:8080/reset-password/21352db78fb5eac7827087ad54e5e0fd13419291` where `21352db78fb5eac7827087ad54e5e0fd13419291` is the token.

Now send a POST request to `/users/reset-password` with your username (email), token and the new password

sample command:
`curl -d "username=antoniochan030@gmail.com&token=21352db78fb5eac7827087ad54e5e0fd1341929&newPassword=testpassword" -X POST http://localhost:8000/users/reset-password`

Sample response should be something similar to this:
```
{"statusCode":200,"statusText":"SUCCESS","data":null}
```

To test that the password was updated, let's try to login with the new password:
- Endpoint to login /users/login:
    -  method: POST
    -  body parameters:
        - `username` type: string
        - `password` type: string

Sample command:
`curl -d "username=antoniochan030@gmail.com&password=testpassword" -X POST http://localhost:8000/users/login`

The response should be :
```
{"statusCode":200,"statusText":"SUCCESS","data":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFudG9uaW9AZ21haWwuY29tbCIsImlhdCI6MTU2NzE5MjA4MCwiZXhwIjoxNTY3Mjc4NDgwfQ.KKh-7r2hpNHEbFD8RTQgMsucjGwGuyKaY7GXAfL8sgI"}
```

Now let's try with an invalid token ie: token=invalidtoken123. We coud also use the same token that was sent to our email. The expiry time should have been updated and the token should be expired after the user successfuly updated their password

sample:
`curl -d "username=antoniochan030@gmail.com&token=invalidToken&newPassword=testpassword" -X POST http://localhost:8000/users/reset-password`

Response should be;
```
{"statusCode":401,"statusText":"UNAUTHORIZED","message":"The authorization credentials provided are insufficient."}
```

There was also another endpoint that was updated due to a refactoring of the getUser function.

- Endpoint to get the user and check if it exists in our DB: `/users/:username`
    -  method: GET

Sample command:
`curl http://localhost:8000/users/antoniochan030@gmail.com`

Sample response:
```
{"statusCode":200,"statusText":"SUCCESS","data":"antoniochan030@gmail.com"}
```

Another sample command use for when the user does not exist:
`curl http://localhost:8000/users/antoniochasdadadan030@gmail.com`

Response:
```
{"statusCode":404,"statusText":"NOT_FOUND","message":"The requested resource or endpoint was not found."}
```

# Next to do
Work on the front-end component that will use these end points.Ticket: https://antoniolok.atlassian.net/browse/SSF-9

howardhenry commented on src/views/routes/users.js:
```diff
@@ -10,13 +12,16 @@ const {
     conflict,
+    notFound,
     serverError,
+    unauthorized,
```
👍 alphabetized

AntonioLok replied:
Haha, learned from many of your previous comments ;)

howardhenry commented on src/views/routes/users.js:
```diff
@@ -61,16 +66,66 @@ router.post('/login', async (req, res) => {
+    const { username } = req.params;
+    await getUser({ username });
+
+    responseHandler.handleSuccess(res, success.CODE);
```
Why does `GET /users/:username` simply return 200? I would expect that some data is returned to the client.

AntonioLok replied:
At first, I was returning the username/email but I decided to not return any data to the client as it's not "needed" by my front-end component. Currently, the client side of my application does not use the data that we are getting back fro this end-point. However, I will include it for future use. But as I said, the data is not being used or expected by my front-end application

AntonioLok replied:
Updated!

howardhenry commented on src/views/routes/users.js:
```diff
@@ -61,16 +66,66 @@ router.post('/login', async (req, res) => {
+    const { message } = error;
+    errLogger.error(error);
+
+    if (message.includes(AUTH__EMAIL_DOES_NOT_EXIST)) {
```
I would not expect a response to `POST /users/forgot-password` to yield an error message called `AUTH__EMAIL_DOES_NOT_EXIST`. Instead, I would expect something that indicates `AUTH__USER_DOES_NOT_EXIST`.

AntonioLok replied:
Not sure what I was thinking when I created this error message *face palm*. Does not even make sense once I have re-read it. Thanks for the suggestion!

howardhenry commented on src/controllers/user/send-recovery-email.js:
```diff
@@ -0,0 +1,44 @@
+    await getUser({ username });
+
+    const token = crypto.randomBytes(20).toString('hex');
+    const { EMAIL_ADDRESS, EMAIL_PASSWORD } = process.env;
```
`EMAIL_ADDRESS` and `EMAIL_PASSWORD` are too ambiguous as environment variables. From what I can see, these are used as the source email address from which system emails are sent. Perhaps you could use more descriptive names for these vars.

AntonioLok replied:
You are right. I have updated these variable names to provide better description.

howardhenry commented on src/views/routes/users.js:
```diff
@@ -61,16 +66,66 @@ router.post('/login', async (req, res) => {
   try {
-    const { email } = req.params;
-    const user = await getUser(email);
+    const { token, newPassword } = req.body;
```
Although the token is random, an attacker with sufficient resources could spam this endpoint until he guesses a valid token, which would effectively reset a user's password.

You can bolster the security of this endpoint by requiring the username to be included in the request body. This way, the token and the matching username would have to be included with the new password in order to reset that user's password.

howardhenry commented on src/controllers/user/get-user.js:
```diff
@@ -1,8 +1,23 @@
+  PASSWORD_RESET_LINK_INVALID_OR_EXPIRED,
+} = constants.errorCodes;
+
+const getUser = async (...fields) => {
```
Logic to find the user from the database should be **fully** abstracted to this controller.  Here you're simply passing in all the fields and only mapping error code logic if `User.getUser` fails. This defeats the purpose of abstraction, which is to move re-used (or potentially re-used) implementation logic to a central place. Doing so also allows you to easily update the database layer without having to refactor the implementation of all the modules that use this abstraction.

AntonioLok replied:
Sorry, I am been having a hard time trying to understand what you are suggesting here. 

Based on my understanding, are you suggesting that I should move the entire logic for getting the user (from this controller) to https://github.com/AntonioLok/online-shopping-reactJS-Backend/blob/master/src/models/user.js#L24?

howardhenry replied:
What you've created here is, in essence, a wrapper for `User.getUser`. It's a function that takes an arbitrary number of fields and queries the user object based on them. This heavily relies on the implementation of Mongo/Mongoose as is not really an abstraction but instead just a wrapper to handle/throw certain error codes.

Instead:
1. think about your use cases
2. create specific functions for each of them
3. if possible, find ways to combine those functions to make them reusable

howardhenry commented on src/controllers/user/update-user.js:
```diff
@@ -0,0 +1,16 @@
+const User = require('../../models/user');
+
+const updateUser = async (username, ...fields) => {
```
I would recommend having one endpoint to update the user (when necessary), then another to change the user's password. There is sufficient logic involved with changing a password that could warrant separating this to a new endpoint. Otherwise, you would end up with a very large function that handles all the nuances of changing a password as well as changing other user properties.

AntonioLok replied:
Good to know! I originally had it as separate functions but thought it was a better idea to have it as a single function as it would reduce the number of functions I have created. But you raised a good point! I will update

howardhenry commented on src/models/user.js:
```diff
@@ -4,6 +4,8 @@ const bcrypt = require('bcryptjs');
   username: { type: String, required: true, index: { unique: true, dropDups: true } },
   password: { type: String, required: true },
+  resetPasswordToken: { type: String },
+  resetPasswordExpires: { type: Number },
```
As this is an expiration, a date would be more appropriate for this property's schema type. This type of better optimized for indexing and will enable you to enter a human readable ISO date when setting the expiration of a field.

AntonioLok commented on package.json:
```diff
@@ -19,13 +19,15 @@
   "dependencies": {
     "bcryptjs": "2.4.3",
     "body-parser": "1.18.3",
+    "crypto": "1.0.1",
```
Ah, I wasn't aware it was deprecated. Good catch!

AntonioLok commented on src/controllers/user/get-user.js:
```diff
@@ -1,8 +1,23 @@
-const getUser = async (email) => {
+const {
+  AUTH__EMAIL_DOES_NOT_EXIST,
+  PASSWORD_RESET_LINK_INVALID_OR_EXPIRED,
```
Done!

AntonioLok commented on src/settings/index.js:
```diff
@@ -0,0 +1,9 @@
+  WEB_APP_BASE_URL: 'https://shopping-site.antoniolok.com',
+};
+
+module.exports = process.env.NODE_ENV === 'production' ? production.WEB_APP_BASE_URL : development.WEB_APP_BASE_URL;
```
I have refactored this. Please take a look at the updated settings file!

howardhenry commented:
Changes look good so far. I'll do a full review again once you're finished.

AntonioLok commented:
Yup. I am just responding to each comments as I go through them. I will let
you know when it should be ready for a full review. Thanks a lot for your
help/time!

On Sat, Dec 28, 2019 at 7:18 PM Howard Henry <notifications@github.com>
wrote:

> Changes look good so far. I'll do a full review again once you're finished.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AntonioLok/online-shopping-reactJS-Backend/pull/20?email_source=notifications&email_token=ADQCPMDHD2DNAIIKSOIPE5LQ27UG7A5CNFSM4J6N24VKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYU3PY#issuecomment-569462207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADQCPMDFL6FMENYDMEY3373Q27UG7ANCNFSM4J6N24VA>
> .
>

AntonioLok commented:
@howardhenry I have addressed all comments. Please kindly review my updated PR whenever you are free! Thanks in advance.

AntonioLok commented on src/views/routes/users.js:
```diff
@@ -61,16 +67,60 @@ router.post('/login', async (req, res) => {
+    const { username } = req.params;
+    const user = await getUser(username);
+
+    responseHandler.handleSuccess(res, success.CODE, user.username);
```
are you proposing to send to the client a user object with the username as its only field? I rather not send the whole user object as it contains private data which I don't want to expose via API calls (such as their password, the token to reset their password, etc).

howardhenry replied:
Ye you definitely don't want to send the entire use object here as it contains private data. You do however want to send back _some_ object i.e. the user object without sensitive data.

Check out the `.select()` method in mongoose which allows you to explicitly include or exclude certain fields when querying an object.

AntonioLok replied:
Sounds good. Thanks for the reply 👍

AntonioLok commented on src/controllers/user/validate-reset-password-token.js:
```diff
@@ -0,0 +1,24 @@
+  AUTH__PASSWORD_RESET_LINK_INVALID_OR_EXPIRED,
+} = constants.errorCodes;
+
+const validateResetPasswordToken = async (username, resetPasswordToken, currentDate) => {
```
👍👍

AntonioLok commented on src/controllers/user/change-user-password.js:
```diff
@@ -0,0 +1,12 @@
+
+const changeUserPassword = async (username, password, resetPasswordExpires) => {
+  try {
+    const hashedPassword = await User.generatePasswordHash(password);
```
Actually, I like this recommendation. I was never aware of this pre/post hooks. I did learn new something new about mongoose thanks to this comment!

AntonioLok commented:
> Changes look good!
> 
> I've added just a few new comments

I have addressed your comments. Kindly review the PR again :D. Btw.... feedback regarding testing instructions are appreciated. Not fully sure if they are really clear to you or helpful when testing. I am open for feedback on how to improve them to make your life easier when reviewing

AntonioLok commented:
> The changes look great!
> 
> I'm so sorry I forgot the notification where you responded that you had made the changes 9 days ago!

Hey, no worries! I just thought you were busy. Thanks for taking the time to review it :). Now I can start working on the next PR :)
